NET-1133: Update config wizard to create v1.0 configs

mondoreale opened the pull request:
## Summary

New Streamr config wizard! 🧙🏻

## Changes

The way the wizard is built is different. I still use the `inquirer` library buy flavoured slightly different. It's more about individual prompt methods (`confirm`, `select`, etc.).

I also test it different. `@inquirer/testing` helped me make the tests more "integrational". I pretend to be the user and validate the output of each scenario.

## How to run locally for a quick test?

```
npm run clean
npm run bootstrap
cd packages/broker
npx esbuild
npx esbuild bin/config-wizard.ts --platform=node --target=node18 --bundle --packages=external > dist/wizard.js && node dist/wizard.js
```

## Example output
```
✓ Congratulations, you've set up your Streamr Network node!
  Your node address is 0x394c8d686B6B083b9bd91DB9f6238140D961e922
  Your node's generated name is Deer Good Foam

  You can start your Streamr node now with
  streamr-broker /Users/mariusz/.streamr/config/default.json

  For environment specific run instructions, see
  https://docs.streamr.network/guides/how-to-run-streamr-node
```

![example](https://github.com/streamr-dev/network/assets/320066/91216d59-1682-4b49-ba9d-875bc6373837)

mondoreale committed:
Shake off

mondoreale committed:
Clean-up the config wizard

mondoreale committed:
Rearrange

mondoreale committed:
Rename

mondoreale committed:
Add more tests

mondoreale committed:
Tweaks, copy, etc.

fonty1 commented:
Generated config file:
<img width="464" alt="Screenshot 2023-12-12 at 12 52 47" src="https://github.com/streamr-dev/network/assets/1593398/789f4b47-7cb3-4f3f-81ef-2719c95244be">
It shouldn't have this line:
❌    "$schema": "https://schema.streamr.network/config-v2.schema.json",
I.e. it should match this template:
https://docs.streamr.network/guides/become-an-operator#testnet-node-config

mondoreale committed:
Eliminate magic

Originally I've made a custom matcher for the answers. The new approach is to generate a `Scenario` object with answers and logs.

This will make it possible to run multiple scenarios within a single test, too.

mondoreale committed:
Add the welcome message; tweak the logger

mondoreale committed:
Tweaks

More empty lines and consistency.

mondoreale committed:
Drop $schema

Copy Pasté… Llooks like some obsolete stuff.

mondoreale commented:
Looks like `npx streamr-broker-init` does it too (copied it over). Gone in [e82b4da](https://github.com/streamr-dev/network/pull/2199/commits/e82b4dab8a16e55a79764d463fbda919ae96761f).

mondoreale committed:
More hand holding

mondoreale committed:
Show balance and nodes loading

fonty1 reviewed (approved):
LGTM. Probably best that @teogeb casts an eye before merging.

mondoreale commented:
Current state.

https://github.com/streamr-dev/network/assets/320066/0d4d4af8-fb65-42a2-a17b-46e33eb4a8f2

Open for suggestions.

mondoreale committed:
Simplify logging

I make a simple markdown-like RegExp-driven text styler to improve readability of this puppy.

mondoreale committed:
Update tests

mondoreale commented:
Pre-approved version (0):

![example](https://github.com/streamr-dev/network/assets/320066/f8676a4c-e5e4-42e5-9ddb-3a36c79cf03f)

mondoreale committed:
Eslint

mondoreale committed:
Fix tests

mondoreale committed:
Fix tests better

This explains the issue.

teogeb commented:
> It shouldn't have this line: ❌ "$schema": "https://schema.streamr.network/config-v2.schema.json", I.e. it should match this template: https://docs.streamr.network/guides/become-an-operator#testnet-node-config

It is good that the schema definition is there as it tells the version number of the config file and therefore allows Broker's auto-migration to be used.

Brokers work ok without the definition. For testnets the version number is not relevant as we don't plan to implement auto-migration between testnets.

mondoreale commented:
@teogeb but doesn't it prevent the client from getting up? Having `$schema` in it constitutes an invalid structure from the schema validator's point of view, no?

mondoreale commented:
Also,

```
curl https://schema.streamr.network/config-v2.schema.json
```

gives

```xml
<Error>
<script/>
<Code>AccessDenied</Code>
<Message>Access Denied</Message>
<RequestId>A60029N99F3JGEY2</RequestId>
<HostId>BupW5G6spzJcilmeTJU1tZBzsP2cynaCIlOHMq8oWXkFFLSFINn8F7zUrQp5iR7wxqAyYQ6DExY=</HostId>
</Error>
```

teogeb commented on packages/broker/src/config/migration.ts:
```diff
@@ -144,7 +145,7 @@ export const readConfigAndMigrateIfNeeded = (fileName: string | undefined): Conf
     if (fileName === undefined) {
         const defaultTargetFile = getDefaultFile()
-        const legacyTargetFile = getLegacyDefaultFile()
+        const legacyTargetFile = path.join(os.homedir(), '.streamr/broker-config.json')
```
Why `getLegacyDefaultFile()` was inlined? Or did the functionality change?

mondoreale replied:
It used to be used once, here, and assigned to `legacyTargetFile`. Does not seem to _need_ a utility function per se, no?

teogeb replied:
Nothing _needs_ a utility function... But we put that into a utility function so that config-related facts are in `config.ts`. If a function which gives us the default file is there, then a function that does almost the same could be also in the same place.

Maybe the actual solution would be to drop the support for legacy file location. The legacy location was used in back in Brubeck, so shouldn't be relevant in Tatum era 👍

mondoreale replied:
The util is back.

mondoreale committed:
Revert "Drop $schema"

This reverts commit e82b4dab8a16e55a79764d463fbda919ae96761f.

teogeb commented on packages/broker/src/config/config.ts:
```diff
@@ -30,20 +30,11 @@ export interface ConfigFile extends Config {
 }
 
-export interface ConfigFile extends Config {
-    $schema?: string
```
Revert this as `$schema` should be used (so that auto-migration works).

mondoreale replied:
Now I see why am I getting über-confused by this, haha. `ConfigFile` got de-dupped here.

![Screenshot 2023-12-13 at 12 44 06 PM](https://github.com/streamr-dev/network/assets/320066/819ed18e-6ce2-4823-890b-fe1e33aab73a)

mondoreale commented:
> @teogeb but doesn't it prevent the client from getting up? Having `$schema` in it constitutes an invalid structure from the schema validator's point of view, no?

Oh, it doesn't fail, eh? I thought @fonty1's note came up from a schema validation failure.

Sounds like it's a good idea to bring it back. @fonty1?

teogeb commented on packages/broker/bin/config-wizard.ts:
```diff
@@ -8,4 +8,10 @@ program
     .description('Run the configuration wizard for the broker')
 
-start()
+void (async () => {
```
```suggestion
;(async () => {
```

mondoreale replied:
`void` is underrated. #nosemis

teogeb replied:
What is the benefit of `void` here? Should we start to use this patten in `network` monorepo codebase?

The `void` is so underrated, that maybe most JS developers (me included) don't even know what this does without guessing / reading the documentation. That's one aspect why this adds unnecessary complexity.

The documentation says:
The `void` operator evaluates the given expression and then returns `undefined` (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/void)

Here the intention is not to evaluate any expression. Of course calling a function technically "evaluates" it as every function returns a value. But here the point is just to execute the function.

mondoreale replied:
Now it's an awkward argument. Let's have it your way. Doesn't make any difference here anyhway.

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,539 @@
-    HTTP: 'http'
+type PubsubPluginKey = 'websocket' | 'mqtt' | 'http'
+
+const DefaultPort: Record<PubsubPluginKey, number> = {
```
```suggestion
const DEFAULT_PORTS: Record<PubsubPluginKey, number> = {
```

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,539 @@
+/**
+ * Lets the user select and configure desired pub/sub plugins.
+ */
+async function getPubsubPlugins() {
```
Could add return type for all functions in this file

fonty1 commented:
> > @teogeb but doesn't it prevent the client from getting up? Having `$schema` in it constitutes an invalid structure from the schema validator's point of view, no?
> 
> Oh, it doesn't fail, eh? I thought @fonty1's note came up from a schema validation failure.
> 
> Sounds like it's a good idea to bring it back. @fonty1?

I assumed it was dropped in 1.0 since that what we've been doing to date :)

> Brokers work ok without the definition. For testnets the version number is not relevant as we don't plan to implement auto-migration between testnets.

New information to me, yeah, follow @teogeb's lead and I'll make a note of this in our Docs as well. Thanks!

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,539 @@
+
+        const storagePath = await getStoragePath()
+
+        const httpServer = http?.port ? { port: http.port } : void 0
```
```suggestion
        const httpServer = http?.port ? { port: http.port } : undefined
```

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,539 @@
+            Object.assign(pubsubPlugins, {
+                http: {},
+            })
+        }
```
Why is this needed?

mondoreale replied:
I'mma port the comments from the OG implementation…

fonty1 replied:
IIRC and IIUC :P including a plugin with default settings is done by adding an empty object like `http: {}` to the plugins section. I can't recall the exact reasoning behind it all though, perhaps we can look at documenting it together when Eric gets back?

mondoreale replied:
This is the only comment that's out there. Doesn't really explain anything.
https://github.com/streamr-dev/network/blob/streamr-1.0/packages/broker/src/config/ConfigWizard.ts#L187

On Slack I described the outcome of this behaviour.

teogeb replied:
The comment is very bad, agree. The summary of Slack discussion:
- for each plugin there should be a block inside plugins block, i.e. if `httpPlugin` enabled, then there is a `http` block
- typically the non-default port is defined inside the plugin's block
  - if a default is used, no port information need to be written as we'll use defaults from the plugin schema, e.g. https://github.com/streamr-dev/network/blob/e00d704440bc68d4455c8a614442c310db9684b4/packages/broker/src/plugins/mqtt/config.schema.json#L11
- the only exception is `httpPlugin`: in that case the port is defined in the `httpServer` root level block
 - if we use default port, then no need to define the root level `httpServer` block (as empty is the default value for it in the Broker schema, and that way it uses the default port automatically) https://github.com/streamr-dev/network/blob/e00d704440bc68d4455c8a614442c310db9684b4/packages/broker/src/config/config.schema.json#L59-L60

teogeb replied:
Maybe we wouldn’t need any exceptional cases at lines `50`-`56`, i.e. we’d have just
```ts
const pubsubPlugins = await getPubsubPlugins()
```
and then when we write pubsubPlugins in plugins, we’d omit the `httpPlugin`’s port:
```ts
plugins: {
    ...operatorPlugins.map((p) => p === ‘http’ ? omit(p, ’port) : p),
    ...pubsubPlugins,
},
```
That way it would a bit clearer that this whole thing is about the port, and if getPubsubPlugins start to give other options in the future, we don’t accidentally omit all httpPlugin options at line 54 (where we currently

mondoreale replied:
That `map` is ugly af, lol. Plus, `pubsubPlugins` is an object. I got your point about the other fields we potentially lose here though. I'll get that fixed.

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,539 @@
+/**
+ * Adjusts the given config for the Mumbai test environment.
+ */
+function getMumbaiConfig(config: ConfigFile): ConfigFile {
```
We can use the new environment preset, i.e. just defined
```
    environment: 'mumbai'
```

https://github.com/streamr-dev/network/pull/2139/files

mondoreale replied:
Very handy! Let's make proper adjustments once the env feature is live.

teogeb replied:
It will be live in the same release as this config wizard

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,539 @@
+
+        persistConfig(
+            storagePath,
+            network === 'polygon' ? config : getMumbaiConfig(config)
```
Could introduce a string literal enum for networks, and also convert this ternary something like `if (network === 'mumbai') { config = getMumbaiConfig(config) }` so that possible third network will be handled correctly (if we add that at some point)

mondoreale replied:
I'll reformat, but a union takes us just as far, currently. I'll keep it simple.

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,539 @@
+            }
+        }
+
+        if (operator) {
```
Coudl use explicit `undefined` checks here and in other if statements

mondoreale replied:
`operator` is either `undefined` or a valid ethereum address.

teogeb replied:
Yes, therefore `if (operator !== undefined)` would communicate that check clearly

mondoreale replied:
It already covers the case and is sufficiently obvious. I'll leave it.

mondoreale commented:
I bring `$schema` back in [19d1058](https://github.com/streamr-dev/network/pull/2199/commits/19d1058fddabee3076db89d321342d89be4eb1f1).

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,544 @@
+                style(`> Your node address has *${f} MATIC* _– checking balance…_`)
+            )
+
+            try {
```
Why is this additional formating needed? Is it for removing `ETH` word?

mondoreale replied:
`progress` prints out a busyness indicator while we're checking the wallet balance. This entire line gets erased and replaced with `Your node address has *12.34 MATIC* (…)` once we know the value.

Both texts are formatted the same way and emphasise the important part.

teogeb replied:
Refered to `.replace(/\.(\d+)/, (f) => f.substring(0, 3))` snippet (maybe that line was the line `92` at the time of comment, or something)

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,544 @@
+            > You can start your Streamr node now with
+            > *streamr-broker ${storagePath}*
+            >
+            > For environment specific run instructions, see
```
Which error case this handles ? E.g. ctrl+c or something else?

mondoreale replied:
Ctrl+c, yes.

teogeb replied:
Does the throw error /force closed/ come from the inquired library? Could add a comment there

(Referring to line `172` here, line numbers corrupted again, maybe because of the force push or something)

mondoreale replied:
A comment here (there) is a decent idea.

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,544 @@
+                ]),
+            }),
+        })
+        .parse(await resp.json())
```
Does this fix some issue with `inquirer/prompts` library?

mondoreale replied:
What issues?

teogeb replied:
Ah time line numbers have changed. Maybe referred the `progress` function.

teogeb reviewed (commented):
First round of comments

teogeb commented:
> @teogeb but doesn't it prevent the client from getting up? Having `$schema` in it constitutes an invalid structure from the schema validator's point of view, no?

This is a Broker config, not a client config. In Broker there is `$schema` property at top level of the config: https://github.com/streamr-dev/network/blob/c09c373a1585c7d3df01aecf7f566802fc8f1e00/packages/broker/src/config/config.schema.json#L8

teogeb commented:
> schema validation failure.

Yes https://linear.app/streamr/issue/NET-1029/v2-broker-config-schema-is-not-accessible-online

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,544 @@
-        auth: {}
-    },
-    plugins: {},
-    apiAuthentication: {
```
Keep the `apiAuthentication` block. It should contain an auto-generated key if any API plugins (i.e. pubsub plugins) have been enabled.

mondoreale replied:
Good to have you, Sir. I'll revert.

mondoreale committed:
Feedback pass

mondoreale committed:
Bring back api keys; more tests

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,591 @@
+            message: 'Please provide the private key to import',
+            validate(value) {
+                try {
+                    return !!new Wallet(value)
```
Why `!!` is used?

mondoreale replied:
Reformatted.

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,591 @@
+                default: defaultPort.toString(),
+                validate(value) {
+                    try {
+                        return !!z.coerce
```
Why `!!` is used?

mondoreale replied:
Reformatted.

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,591 @@
+/**
+ * Lets the user decide the desired network for their node.
+ */
+async function getNetwork(): Promise<'polygon' | 'mumbai'> {
```
Could introduce an enum for the networks

mondoreale replied:
I'll give it a union type so I don't have to typeof it later. Enum brings no benefit.

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,591 @@
+            })
+        )
+
+        pubsubPlugins[key] = port !== defaultPort ? { port } : {}
```
We check if the port has already been taken at line 349. To support that, maybe it would be good to do the omitting of default ports after all the ports have been asked? 

(And it would maybe make sense to omit just the `port` from the object instead of assuming that the port is the only config option we ask. Not too relevant here, but if we add other config options, it is easier if the port omitting feature doesn't give us a surprise by omitting all options.)

mondoreale replied:
> We check if the port has already been taken at line 349. To support that, maybe it would be good to do the omitting of default ports after all the ports have been asked?

Not sure if I understand you right, but I already prevent conflicts on both custom and default ports. It's also tested.

> (And it would maybe make sense to omit just the `port` from the object instead of assuming that the port is the only config option we ask. Not too relevant here, but if we add other config options, it is easier if the port omitting feature doesn't give us a surprise by omitting all options.)

Let's be today.

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,591 @@
+): Promise<BigNumber> {
+    const url = config[network].rpcEndpoints[0]?.url || ''
+
+    if (!/^https?:/i.test(url)) {
```
```suggestion
    const url = config[network].rpcEndpoints[0]?.url

    if ((url === undefined) || !/^https?:/i.test(url)) {
```

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,591 @@
-        answers = await inquirer.prompt(storagePathPrompts)
-    } while (answers.overwrite === false)
-    return answers as any
+    return data.operator?.nodes || null
```
```suggestion
    return data.operator?.nodes ?? []
```

the return type of this function can be just `string[]` and the check at line `146` be` if (nodes.length > 0)`

mondoreale replied:
What if the `operator` is null? (= meaning not found)

teogeb commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,591 @@
+                    }
+                } else {
+                    log(`
+                        > x Your Operator could not be found on the **${network.replace(/\w/, (l) => l.toUpperCase())}** network, see
```
What the `network.replace(/\w/, (l) => l.toUpperCase())` does? E.g. does it capitalize the string (if that is what it does, could use https://lodash.com/docs/4.17.15#capitalize)

teogeb commented on packages/broker/test/unit/ConfigWizard.test.ts:
```diff
@@ -1,257 +1,1689 @@
+        const log = args
+            .join('')
+            // eslint-disable-next-line no-control-regex
+            .replace(/\x1B\[\d+m/g, '')
```
What this does? (Removes colors, maybe?)

mondoreale committed:
Bring back `getLegacyDefaultFile`

Meh.

mondoreale committed:
Use lodash's `capitalize`

mondoreale committed:
Comment on the forced quits

mondoreale committed:
Scary `void` → conformistic `;`

mondoreale committed:
Give network keys their well-deserved type

mondoreale committed:
Add a short note about `http` plugin's port

mondoreale commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,591 @@
+
+                const content = `Your node address has *${utils
+                    .formatEther(balance)
+                    .replace(/\.(\d+)/, (f) => f.substring(0, 3))} MATIC*`
```
Oh, this? Say you have a balance of `19.7465234786528937`. This makes it `19.74`.

mondoreale committed:
Drop unnecessary rexexp grouping

mondoreale committed:
Use friendlier syntax for them validators

mondoreale committed:
Readability

mondoreale commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,607 @@
+            message: 'Please provide the private key to import',
+            validate(value) {
+                try {
+                    new Wallet(value)
+
+                    return true
```
`!!` gone.

mondoreale commented on packages/broker/src/config/ConfigWizard.ts:
```diff
@@ -1,254 +1,607 @@
+                default: defaultPort.toString(),
+                validate(value) {
+                    try {
+                        z.coerce
```
And here, `!!` gone.

mondoreale committed:
Simplify mumbai configs

We're using the yet to be released (?) feature: environments.

mondoreale commented:
@teogeb the environment field is in. The PR turned a bit chaotic with all the things flying in, and the comments spread out unevenly. I don't remember gh being so… inconsistent about rebasing and comments ¯\\_(ツ)_/¯.

mondoreale committed:
Drop immer

mondoreale committed:
Rename network to environmentId; update types

mondoreale committed:
Simplify balance formatter

mondoreale committed:
Give `progress` a better name; format

mondoreale committed:
Being explicit

mondoreale committed:
Comment on the weird regexp

mondoreale committed:
Update tests

mondoreale committed:
Swap

mondoreale committed:
Test the new wizard

mondoreale commented:
This is now better. I dropped the unexpected `@inquirer/testing` from the root `package.json` and kept the `@streamr/config@5.0.5` update from `streamr-1.0`.
